html/beps/bep_encrypted_data.rst

```diff
@@ -0,0 +1,297 @@
+  The protocol version used to encrypt the torrent, currently *1*. New versions may be introduced by updates to this BEP if cryptographic weaknesses necessitate incompatible changes. Implementations should check if they support the version indicated in the metadata file and otherwise inform the user that they can download the data but not decrypt it.
+
+``shadow``
+  bencoded-then-encrypted dictionary of key-value pairs that shadow entries in the info dictionary.
+  If it is absent only the payload is encrypted and all info dictionary entries are non-shadowed.
+  Implementations should only shadow a whitelist of keys which they have a shadowing strategy and ignore other keys.
+  Shadowable keys suggested by this BEP: ``length, files, name, comment``.
+
+``mac``
+  message authentication code covering the info dictionary
+
+``name``
+  the name field is a mandatory part of [BEP 3]_, therefore a placeholder MUST be provided if a shadow name is used. An implementation may either generate a random string consisting of filesystem-friendly characters or allow the user to choose a public name that reveals less information than the shadow name.
+
+``files``, ``length``
+  The shadow dictionary MAY override the single/multifile nature indicated by the public info dictionary. If it does not shadow either key then the public information is canonical.
+  If the ``files`` or ``length`` are shadowed then the overall payload length MUST be consistent with the public version.
+  If a shadow dictionary is present the public information should be treated as decorative / advisory until it can be determined whether it has been shadowed, i.e. until the shadow data can be decrypted. 
+
+
+To protect privacy an implementation should use shadowing for any additional keys that reveal information about the payload
+
+
+Encryption
+==========
+
+Building blocks used:  SHA2-256[rfc6234]_, ChaCha20[rfc7539]_, HMAC[rfc2104]_, PBKDF2[rfc2898]_  
+
+.. parsed-literal::
+
+    Key.root = random key, recommended strength: 256bits
+
+    Key.payload = PBKDF2(HMAC−SHA256, Key.root, salt || "payload", 4096, 256)
+
+    Key.shadow =  sha256(Key.payload || "shadow")
+    
+    mac = HMAC−SHA256(info-dict without mac, Key.shadow)
+
+    IV.payload = truncate_64(sha256(salt || "payload"))
+
+    IV.shadow = truncate_64(sha256(salt || "shadow"))
+
+PBKDF2 key derivation is used in case root keys with less entropy than the recommended are used, e.g. for password-based schemes. But for general use this BEP assumes that the root key consists of random binary data and hence mandates hexadecimal encoding when the keys need to be displayed in a human-readable format.
+
+ChaCha20 is used to both encrypt the shadow dictionary and the torrent payload.
+
+The optional ``shadow`` dictionary is encrypted after bencoding with ``Key.shadow`` and ``IV.shadow``.
+
+The ``mac`` is calculated over the bencoded info-dictionary including the ``bepXX`` dictionary but excluding the ``mac`` key value pair.
```

Comment:
yeah, I simply assumed two encoding passes. If we want to enable some clever single-pass logic then a placeholder simplifies because it allows in-place editing is easier than splicing.